src/net/server/connection.rs

```diff
@@ -0,0 +1,633 @@
+use core::ops::ControlFlow;
+use core::sync::atomic::Ordering;
+use core::time::Duration;
+use std::io;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use tokio::io::{
+    AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt, ReadHalf, WriteHalf,
+};
+use tokio::sync::mpsc::Sender;
+use tokio::sync::{mpsc, watch};
+use tokio::task::JoinHandle;
+use tracing::{debug, error};
+
+//------------ Connection -----------------------------------------------
+
+pub struct Connection<Stream, Buf, Svc>
+where
+    Stream: AsyncRead + AsyncWrite + Send + Sync + 'static,
+    Buf: BufSource + Send + Sync + 'static + Clone,
+    Buf::Output: Send + Sync + 'static,
+    Svc: Service<Buf::Output> + Send + Sync + 'static + Clone,
+{
+    buf_source: Buf,
+    metrics: Arc<ServerMetrics>,
+    service: Svc,
+    middleware_chain: Option<MiddlewareChain<Buf::Output, Svc::Target>>,
+    stream: Option<Stream>,
+    addr: SocketAddr,
+}
+
+/// Creation
+///
+impl<Stream, Buf, Svc> Connection<Stream, Buf, Svc>
+where
+    Stream: AsyncRead + AsyncWrite + Send + Sync + 'static,
+    Buf: BufSource + Send + Sync + 'static + Clone,
+    Buf::Output: Send + Sync + 'static,
+    Svc: Service<Buf::Output> + Send + Sync + 'static + Clone,
+{
+    #[must_use]
+    pub fn new(
+        service: Svc,
+        middleware_chain: Option<MiddlewareChain<Buf::Output, Svc::Target>>,
+        buf_source: Buf,
+        metrics: Arc<ServerMetrics>,
+        stream: Stream,
+        addr: SocketAddr,
+    ) -> Self {
+        Self {
+            buf_source,
+            service,
+            middleware_chain,
+            metrics,
+            stream: Some(stream),
+            addr,
+        }
+    }
+}
+
+/// Control
+///
+impl<Stream, Buf, Svc> Connection<Stream, Buf, Svc>
+where
+    Stream: AsyncRead + AsyncWrite + Send + Sync + 'static,
+    Buf: BufSource + Send + Sync + 'static + Clone,
+    Buf::Output: Send + Sync + 'static,
+    Svc: Service<Buf::Output> + Send + Sync + 'static + Clone,
+{
+    /// Start reading requests and writing responses to the stream.
+    ///
+    /// # Shutdown behaviour
+    ///
+    /// When the parent server is shutdown (explicitly or via Drop) the child
+    /// connections will also see the [`ServiceCommand::Shutdown`] signal and
+    /// shutdown and flush any pending writes to the output stream.
+    ///
+    /// Any requests received after the shutdown signal or requests still
+    /// in-flight will be abandoned.
+    ///
+    /// TODO: What does "abandoned" mean in practice here?
+    pub async fn run(mut self, command_rx: watch::Receiver<ServiceCommand>)
+    where
+        Svc::Single: Send,
+    {
+        self.metrics
+            .num_connections
+            .as_ref()
+            .unwrap()
+            .fetch_add(1, Ordering::Relaxed);
+
+        // TODO: Why use an Option and then Option::take()?
+        let stream = self.stream.take().unwrap();
+        self.run_until_error(command_rx, stream).await;
+
+        self.metrics
```

Comment:
Good catch. It might be better to move these metric inc/dec calls into the `run_until_error()` code tree so that their execution is tied to the lifetime of the code they refer to, and thus also avoid the problem you identified.

Reply:
Hmm that won't help because then it just has the same problem at some other `.await` point. Indeed it needs a Drop impl.